ledger/src/blockstore.rs

```diff
@@ -1967,14 +1967,18 @@ impl Blockstore {
         self.blocktime_cf.get(slot)
     }
 
-    pub fn get_rooted_block_time(&self, slot: Slot) -> Result<Option<UnixTimestamp>> {
+    pub fn get_rooted_block_time(&self, slot: Slot) -> Result<UnixTimestamp> {
         datapoint_info!(
             "blockstore-rpc-api",
             ("method", "get_rooted_block_time", String)
         );
         let _lock = self.check_lowest_cleanup_slot(slot)?;
+
         if self.is_root(slot) {
-            return self.blocktime_cf.get(slot);
+            return self
+                .blocktime_cf
+                .get(slot)?
+                .ok_or(BlockstoreError::SlotUnavailable);
```

Comment:
I was a little on the fence about what error to return here.

At this point, we know the slot was rooted, and that the slot was present in roots column. Thus, if we get a `None` from the blocktime column, that would point to an inconsistency between the two columns. This would either be from:
- Blocktime somehow not getting inserted for this slot OR
- This slot getting purged from blocktime column but not roots column somehow (recall that LedgerCleanupService is atomic for slot ranges on all columns so this case really isn't possible)

Neither of these should happen ever. I landed on `SlotUnavailable` given that the error variant already exists, but something like `InternalError` or `ColumnDataInconsistency` also came to mind. For completeness, `.unwrap()` or `.expect()` could arguably fit here too, but I typically try to use those an minimally as possible incase there is some corner case that we're overlooking

Reply:
> LedgerCleanupService is atomic for slot ranges on all columns
I think technically this isn't true for CompactionFilter columns. On a ledger-cleanup iteration, I think all that happens for these is that oldest_slot gets moved forward and the older slot data are eligible for cleanup next periodic compaction.

Since we're calling `check_lowest_cleanup_slot()` here, the fact that older data may still be hanging around shouldn't be relevant. Mostly just articulating this to help us remember to check the cleanup slot when doing stuff with those columns.
...or maybe we should have some special getter function for these columns that checks the oldest_slot and refuses to return older data?

To your actual topic here, I bet there are other places we should be using an `InternalError` or `ColumnDataInconsistency` variant 😅 . But I'm fine with `SlotUnavailable`. As you say, we really don't expect this to ever be returned.

Reply:
> I think technically this isn't true for CompactionFilter columns. On a ledger-cleanup iteration, I think all that happens for these is that oldest_slot gets moved forward and the older slot data are eligible for cleanup next periodic compaction.

True. I guess I meant (but failed to indicate) that I was referring to the slot based columns that are range-deleted here. In this case, blocktime and root are both slot-based:
https://github.com/solana-labs/solana/blob/6624a09d381dde44bc6abf06e2145fc25586631e/ledger/src/blockstore/blockstore_purge.rs#L164-L223

But yes, you are correct that the "special" columns are available until periodic compactions cleans them up.

> Mostly just articulating this to help us remember to check the cleanup slot when doing stuff with those columns.
...or maybe we should have some special getter function for these columns that checks the oldest_slot and refuses to return older data?

Definitely something worth considering. The "higher" level functions (ie the ones that are called by `rpc.rs`) have the `check_lowest_cleanup_slot()` call as you mentioned, but that isn't across the board. So, we currently have to remember to do the check everywhere we want it. Will mull on this a bit. The competing argument is to minimize overhead; as my other PR showed, it is easy to get some redundant (and non-trivial) checks when we start stacking functions up.

> To your actual topic here, I bet there are other places we should be using an InternalError or ColumnDataInconsistency variant 😅 . But I'm fine with SlotUnavailable. As you say, we really don't expect this to ever be returned.

Cool, maybe a cleanup task for The Future™️

Reply:
> In this case, blocktime and root are both slot-based

Oh yes, true! Sorry for the non sequitur!